plutus-contract/src/Plutus/Contract/StateMachine.hs

```diff
@@ -230,66 +249,83 @@ waitForUpdateUntilSlot ::
-    let states = txns >>= getStates scInstance . outputsMapFromTxForAddress addr
-    case states of
-        [] | slot < timeoutSlot -> pure ContractEnded
-        [] | slot >= timeoutSlot -> pure $ Timeout timeoutSlot
-        xs -> case scChooser xs of
-                Left err         -> throwing _SMContractError err
-                Right (state, _) -> pure $ WaitingResult (tyTxOutData state)
+    -> Contract w schema e (WaitingResult Slot i state)
+waitForUpdateUntilSlot client timeoutSlot = waitForUpdateTimeout client (isSlot timeoutSlot) >>= fmap (fmap (tyTxOutData . ocsTxOut)) . awaitPromise
 
 -- | Same as 'waitForUpdateUntilSlot', but works with 'POSIXTime' instead.
 waitForUpdateUntilTime ::
     ( AsSMContractError e
     , AsContractError e
     , PlutusTx.FromData state
     , PlutusTx.ToData state
+    , PlutusTx.FromData i
     )
     => StateMachineClient state i
     -> POSIXTime
-    -> Contract w schema e (WaitingResult state)
-waitForUpdateUntilTime sm timeoutTime = do
-    waitForUpdateUntilSlot sm $ TimeSlot.posixTimeToEnclosingSlot def timeoutTime
+    -> Contract w schema e (WaitingResult Slot i state)
+waitForUpdateUntilTime sm timeoutTime = waitForUpdateUntilSlot sm (TimeSlot.posixTimeToEnclosingSlot def timeoutTime)
 
 -- | Wait until the on-chain state of the state machine instance has changed,
 --   and return the new state, or return 'Nothing' if the instance has been
 --   terminated. If 'waitForUpdate' is called before the instance has even
 --   started then it returns the first state of the instance as soon as it
 --   has started.
 waitForUpdate ::
+    forall state i w schema e.
     ( AsSMContractError e
     , AsContractError e
     , PlutusTx.FromData state
     , PlutusTx.ToData state
+    , PlutusTx.FromData i
     )
     => StateMachineClient state i
     -> Contract w schema e (Maybe (OnChainState state i))
-waitForUpdate StateMachineClient{scInstance, scChooser} = do
-    let addr = Scripts.validatorAddress $ typedValidator scInstance
-    txns <- nextTransactionsAt addr
-    let states = txns >>= getStates scInstance . outputsMapFromTxForAddress addr
-    case states of
-        [] -> pure Nothing
-        xs -> either (throwing _SMContractError) (pure . Just) (scChooser xs)
+waitForUpdate client = waitForUpdateTimeout client never >>= awaitPromise >>= \case
+    Timeout t        -> absurd t
+    ContractEnded{}  -> pure Nothing
+    InitialState _ r -> pure (Just r)
+    Transition _ _ r -> pure (Just r)
+
+-- | Wait until the on-chain state of the state machine changes, or a timeout.
+waitForUpdateTimeout ::
+    forall state i t w schema e.
+    ( AsSMContractError e
+    , AsContractError e
+    , PlutusTx.FromData state
+    , PlutusTx.ToData state
+    , PlutusTx.FromData i
+    )
+    => StateMachineClient state i -- ^ The state machine client
+    -> Promise w schema e t -- ^ The timeout
+    -> Contract w schema e (Promise w schema e (WaitingResult t i (OnChainState state i)))
```

Comment:
True, but at least the signature tells you that it has to run some `Contract` effects before it can construct the `Promise`. I changed the comment say that.